File: packages/kotlin/reader-ui-runtime/src/main/kotlin/io/reader/ui/runtime/ReaderUITypedPayloadValidator.kt
package io.reader.ui.runtime

import java.math.BigDecimal
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.intOrNull

private val readerUISafeIntegerLimit = BigDecimal("9007199254740991")

fun validateReaderUITypedPayload(
    event: String,
    payload: ReaderUIJSONPayload
): ReaderUITypedPayloadContract? {
    val contract = GeneratedRuntimeTypedPayloadContracts.byEvent[event] ?: return null
    validateTypedPayloadSchema(contract.schema, JsonObject(payload), "payload")
    return contract
}

fun validateReaderUITypedResult(
    event: String,
    effectType: String,
    result: ReaderUIJSONResult
): ReaderUITypedResultContract {
    val contract = GeneratedRuntimeTypedPayloadContracts.byEvent[event]
        ?: throw ReaderUIRuntimeException("INVALID_TYPED_CONTRACT", "No typed contract for $event")
    val resultContract = contract.resultSchemas[effectType]
        ?: throw ReaderUIRuntimeException("UNDECLARED_TYPED_RESULT", "$event does not declare a result for $effectType")
    try {
        validateTypedPayloadSchema(resultContract.schema, JsonObject(result), "result")
    } catch (failure: ReaderUIRuntimeException) {
        if (failure.code == "INVALID_TYPED_PAYLOAD") {
            throw ReaderUIRuntimeException("INVALID_TYPED_RESULT", failure.message ?: "invalid typed result")
        }
        throw failure
    }
    return resultContract
}

private fun invalidTypedPayload(path: String, message: String): Nothing =
    throw ReaderUIRuntimeException("INVALID_TYPED_PAYLOAD", "$path $message")

private fun validateTypedJSONValue(value: JsonElement, path: String) {
    when (value) {
        is JsonObject -> value.forEach { (key, item) -> validateTypedJSONValue(item, "$path.$key") }
        is JsonArray -> value.forEachIndexed { index, item -> validateTypedJSONValue(item, "$path[$index]") }
        is JsonPrimitive -> if (!value.isString) {
            val decimal = value.content.toBigDecimalOrNull() ?: return
            if (decimal.stripTrailingZeros().scale() <= 0 && decimal.abs() > readerUISafeIntegerLimit) {
                invalidTypedPayload(path, "must contain only finite numbers and IEEE-754 safe integers")
            }
        }
        else -> Unit
    }
}

private fun schemaObject(schema: JsonElement, path: String): JsonObject = schema as? JsonObject
    ?: throw ReaderUIRuntimeException("INVALID_TYPED_CONTRACT", "$path schema must be an object")

private fun JsonObject.string(key: String): String? =
    (this[key] as? JsonPrimitive)?.takeIf { it.isString }?.content

private fun validateTypedPayloadSchema(rawSchema: JsonElement, value: JsonElement, path: String) {
    val schema = schemaObject(rawSchema, path)
    if (value === JsonNull) {
        if ((schema["nullable"] as? JsonPrimitive)?.booleanOrNull == true) return
        invalidTypedPayload(path, "must not be null")
    }
    (schema["oneOf"] as? JsonArray)?.let { branches ->
        var accepted = 0
        branches.forEach { branch ->
            try {
                validateTypedPayloadSchema(branch, value, path)
                accepted += 1
            } catch (error: ReaderUIRuntimeException) {
                if (error.code != "INVALID_TYPED_PAYLOAD") throw error
            }
        }
        if (accepted != 1) invalidTypedPayload(path, "must match exactly one schema variant (matched $accepted)")
        return
    }
    val type = schema.string("type")
        ?: throw ReaderUIRuntimeException("INVALID_TYPED_CONTRACT", "$path schema type is missing")
    when (type) {
        "object" -> {
            val target = value as? JsonObject ?: invalidTypedPayload(path, "must be an object")
            val properties = schema["properties"] as? JsonObject ?: JsonObject(emptyMap())
            (schema["required"] as? JsonArray).orEmpty().forEach { item ->
                val field = (item as? JsonPrimitive)?.contentOrNull ?: return@forEach
                if (!target.containsKey(field)) invalidTypedPayload("$path.$field", "is required")
            }
            if ((schema["additionalProperties"] as? JsonPrimitive)?.booleanOrNull == false) {
                target.keys.forEach { field ->
                    if (!properties.containsKey(field)) invalidTypedPayload("$path.$field", "is unknown")
                }
            } else if (schema["additionalProperties"] is JsonObject) {
                val additionalSchema = schema.getValue("additionalProperties")
                target.forEach { (field, child) ->
                    if (!properties.containsKey(field)) validateTypedPayloadSchema(additionalSchema, child, "$path.$field")
                }
            }
            properties.forEach { (field, childSchema) ->
                target[field]?.let { validateTypedPayloadSchema(childSchema, it, "$path.$field") }
            }
            (schema["constraints"] as? JsonArray).orEmpty().forEach { constraintElement ->
                val constraint = schemaObject(constraintElement, path)
                when (constraint.string("kind")) {
                    "nonBlankWhen" -> {
                        val field = constraint.string("field") ?: return@forEach
                        val whenField = constraint.string("whenField") ?: return@forEach
                        val expected = (constraint["equals"] as? JsonPrimitive)?.booleanOrNull
                        if ((target[whenField] as? JsonPrimitive)?.booleanOrNull == expected) {
                            val text = (target[field] as? JsonPrimitive)?.takeIf { it.isString }?.content
                            if (text.isNullOrBlank()) invalidTypedPayload("$path.$field", "must be non-blank for this variant")
                        }
                    }
                    "booleanAnyTrue" -> {
                        val fields = constraint["fields"] as? JsonArray ?: return@forEach
                        val defaults = constraint["defaults"] as? JsonObject ?: JsonObject(emptyMap())
                        val anyTrue = fields.any { item ->
                            val field = (item as? JsonPrimitive)?.contentOrNull ?: return@any false
                            ((target[field] ?: defaults[field]) as? JsonPrimitive)?.booleanOrNull == true
                        }
                        if (!anyTrue) invalidTypedPayload(path, "requires one target scope to be true")
                    }
                    "validRegexWhen" -> {
                        val field = constraint.string("field") ?: return@forEach
                        val flagField = constraint.string("flagField") ?: return@forEach
                        val enabled = (target[flagField] as? JsonPrimitive)?.booleanOrNull
                            ?: (constraint["default"] as? JsonPrimitive)?.booleanOrNull
                            ?: false
                        if (enabled) {
                            val pattern = (target[field] as? JsonPrimitive)?.takeIf { it.isString }?.content ?: return@forEach
                            try { Regex(pattern) } catch (_: IllegalArgumentException) {
                                invalidTypedPayload("$path.$field", "must be a valid regular expression")
                            }
                        }
                    }
                    "stringAnyNonBlank" -> {
                        val fields = constraint["fields"] as? JsonArray ?: return@forEach
                        val anyNonBlank = fields.any { item ->
                            val field = (item as? JsonPrimitive)?.contentOrNull ?: return@any false
                            (target[field] as? JsonPrimitive)?.takeIf { it.isString }?.content?.isNotBlank() == true
                        }
                        if (!anyNonBlank) invalidTypedPayload(path, "requires one string field to be non-blank")
                    }
                    else -> throw ReaderUIRuntimeException("INVALID_TYPED_CONTRACT", "$path has an unsupported constraint")
                }
            }
            return
        }
        "array" -> {
            val array = value as? JsonArray ?: invalidTypedPayload(path, "must be an array")
            val minimum = (schema["minItems"] as? JsonPrimitive)?.intOrNull
            if (minimum != null && array.size < minimum) invalidTypedPayload(path, "must contain at least $minimum items")
            val items = schema["items"]
                ?: throw ReaderUIRuntimeException("INVALID_TYPED_CONTRACT", "$path array schema has no items")
            array.forEachIndexed { index, item -> validateTypedPayloadSchema(items, item, "$path[$index]") }
            return
        }
        "string" -> {
            val string = (value as? JsonPrimitive)?.takeIf { it.isString }?.content
                ?: invalidTypedPayload(path, "must be a string")
            val minimum = (schema["minLength"] as? JsonPrimitive)?.intOrNull
            if (minimum != null && string.length < minimum) invalidTypedPayload(path, "must contain at least $minimum characters")
            val maximum = (schema["maxLength"] as? JsonPrimitive)?.intOrNull
            if (maximum != null && string.length > maximum) invalidTypedPayload(path, "must contain at most $maximum characters")
            if ((schema["nonBlank"] as? JsonPrimitive)?.booleanOrNull == true && string.isBlank()) {
                invalidTypedPayload(path, "must be non-blank")
            }
            schema.string("pattern")?.let { pattern ->
                val regex = try { Regex(pattern) } catch (_: IllegalArgumentException) {
                    throw ReaderUIRuntimeException("INVALID_TYPED_CONTRACT", "$path has an invalid schema pattern")
                }
                if (!regex.containsMatchIn(string)) invalidTypedPayload(path, "does not match its required pattern")
            }
        }
        "integer" -> {
            val primitive = (value as? JsonPrimitive)?.takeUnless { it.isString }
                ?: invalidTypedPayload(path, "must be an IEEE-754 safe integer")
            val decimal = primitive.content.toBigDecimalOrNull()
                ?: invalidTypedPayload(path, "must be an IEEE-754 safe integer")
            if (decimal.stripTrailingZeros().scale() > 0 || decimal.abs() > readerUISafeIntegerLimit) {
                invalidTypedPayload(path, "must be an IEEE-754 safe integer")
            }
        }
        "number" -> {
            val number = (value as? JsonPrimitive)?.takeUnless { it.isString }?.doubleOrNull
            if (number == null || !number.isFinite()) invalidTypedPayload(path, "must be a finite number")
        }
        "boolean" -> if ((value as? JsonPrimitive)?.takeUnless { it.isString }?.booleanOrNull == null) {
            invalidTypedPayload(path, "must be a boolean")
        }
        "json" -> {
            validateTypedJSONValue(value, path)
            return
        }
        else -> throw ReaderUIRuntimeException("INVALID_TYPED_CONTRACT", "$path has unsupported schema type $type")
    }

    schema["const"]?.let { if (it != value) invalidTypedPayload(path, "does not equal its required constant") }
    (schema["enum"] as? JsonArray)?.let { if (value !in it) invalidTypedPayload(path, "is not an allowed enum value") }
    val number = (value as? JsonPrimitive)?.takeUnless { it.isString }?.content?.toBigDecimalOrNull()
    val minimum = (schema["minimum"] as? JsonPrimitive)?.content?.toBigDecimalOrNull()
    val maximum = (schema["maximum"] as? JsonPrimitive)?.content?.toBigDecimalOrNull()
    if (number != null && minimum != null && number < minimum) invalidTypedPayload(path, "must be >= $minimum")
    if (number != null && maximum != null && number > maximum) invalidTypedPayload(path, "must be <= $maximum")
}

File: packages/swift/ReaderUIRuntime/Sources/ReaderUIRuntime/ReaderUITypedPayloadValidator.swift
import Foundation

private let readerUISafeIntegerLimit = 9_007_199_254_740_991.0

@discardableResult
public func validateReaderUITypedPayload(
    event: String,
    payload: ReaderUIJSONPayload
) throws -> ReaderUITypedPayloadContract? {
    guard let contract = GeneratedRuntimeTypedPayloadContracts.byEvent[event] else { return nil }
    try validateTypedPayloadSchema(contract.schema, value: .object(payload), path: "payload")
    return contract
}

@discardableResult
public func validateReaderUITypedResult(
    event: String,
    effectType: String,
    result: ReaderUIJSONResult
) throws -> ReaderUITypedResultContract {
    guard let contract = GeneratedRuntimeTypedPayloadContracts.byEvent[event] else {
        throw ReaderUIRuntimeFailure(code: "INVALID_TYPED_CONTRACT", message: "No typed contract for \(event)")
    }
    guard let resultContract = contract.resultSchemas[effectType] else {
        throw ReaderUIRuntimeFailure(
            code: "UNDECLARED_TYPED_RESULT",
            message: "\(event) does not declare a result for \(effectType)"
        )
    }
    do {
        try validateTypedPayloadSchema(resultContract.schema, value: .object(result), path: "result")
    } catch let failure as ReaderUIRuntimeFailure where failure.code == "INVALID_TYPED_PAYLOAD" {
        throw ReaderUIRuntimeFailure(code: "INVALID_TYPED_RESULT", message: failure.message)
    }
    return resultContract
}

private func invalidTypedPayload(_ path: String, _ message: String) -> ReaderUIRuntimeFailure {
    ReaderUIRuntimeFailure(code: "INVALID_TYPED_PAYLOAD", message: "\(path) \(message)")
}

private func validateTypedJSONValue(_ value: ReaderUIJSONValue, path: String) throws {
    switch value {
    case .number(let number):
        guard number.isFinite,
              number.rounded(.towardZero) != number || abs(number) <= readerUISafeIntegerLimit else {
            throw invalidTypedPayload(path, "must contain only finite numbers and IEEE-754 safe integers")
        }
    case .array(let values):
        for (index, item) in values.enumerated() {
            try validateTypedJSONValue(item, path: "\(path)[\(index)]")
        }
    case .object(let object):
        for (key, item) in object {
            try validateTypedJSONValue(item, path: "\(path).\(key)")
        }
    default:
        break
    }
}

private func schemaObject(_ schema: ReaderUIJSONValue, path: String) throws -> ReaderUIJSONPayload {
    guard case .object(let object) = schema else {
        throw ReaderUIRuntimeFailure(code: "INVALID_TYPED_CONTRACT", message: "\(path) schema must be an object")
    }
    return object
}

private func validateTypedPayloadSchema(
    _ rawSchema: ReaderUIJSONValue,
    value: ReaderUIJSONValue,
    path: String
) throws {
    let schema = try schemaObject(rawSchema, path: path)
    if value == .null {
        if schema["nullable"]?.boolValue == true { return }
        throw invalidTypedPayload(path, "must not be null")
    }
    if case .array(let branches)? = schema["oneOf"] {
        var accepted = 0
        for branch in branches {
            do {
                try validateTypedPayloadSchema(branch, value: value, path: path)
                accepted += 1
            } catch let failure as ReaderUIRuntimeFailure where failure.code == "INVALID_TYPED_PAYLOAD" {
                continue
            }
        }
        guard accepted == 1 else {
            throw invalidTypedPayload(path, "must match exactly one schema variant (matched \(accepted))")
        }
        return
    }
    guard let type = schema["type"]?.stringValue else {
        throw ReaderUIRuntimeFailure(code: "INVALID_TYPED_CONTRACT", message: "\(path) schema type is missing")
    }

    switch type {
    case "object":
        guard case .object(let object) = value else { throw invalidTypedPayload(path, "must be an object") }
        let properties: ReaderUIJSONPayload
        if case .object(let declared)? = schema["properties"] { properties = declared } else { properties = [:] }
        if case .array(let required)? = schema["required"] {
            for field in required.compactMap(\.stringValue) where object[field] == nil {
                throw invalidTypedPayload("\(path).\(field)", "is required")
            }
        }
        if schema["additionalProperties"]?.boolValue == false {
            for field in object.keys where properties[field] == nil {
                throw invalidTypedPayload("\(path).\(field)", "is unknown")
            }
        } else if case .object? = schema["additionalProperties"],
                  let additionalSchema = schema["additionalProperties"] {
            for field in object.keys where properties[field] == nil {
                try validateTypedPayloadSchema(additionalSchema, value: object[field]!, path: "\(path).\(field)")
            }
        }
        for (field, childSchema) in properties {
            if let child = object[field] {
                try validateTypedPayloadSchema(childSchema, value: child, path: "\(path).\(field)")
            }
        }
        if case .array(let constraints)? = schema["constraints"] {
            for constraintValue in constraints {
                let constraint = try schemaObject(constraintValue, path: path)
                switch constraint["kind"]?.stringValue {
                case "nonBlankWhen":
                    guard let field = constraint["field"]?.stringValue,
                          let whenField = constraint["whenField"]?.stringValue else { continue }
                    if object[whenField]?.boolValue == constraint["equals"]?.boolValue,
                       object[field]?.stringValue?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty != false {
                        throw invalidTypedPayload("\(path).\(field)", "must be non-blank for this variant")
                    }
                case "booleanAnyTrue":
                    guard case .array(let fields)? = constraint["fields"] else { continue }
                    let defaults: ReaderUIJSONPayload
                    if case .object(let values)? = constraint["defaults"] { defaults = values } else { defaults = [:] }
                    let anyTrue = fields.compactMap(\.stringValue).contains { field in
                        (object[field] ?? defaults[field])?.boolValue == true
                    }
                    if !anyTrue {
                        throw invalidTypedPayload(path, "requires one target scope to be true")
                    }
                case "validRegexWhen":
                    guard let field = constraint["field"]?.stringValue,
                          let flagField = constraint["flagField"]?.stringValue else { continue }
                    let enabled = object[flagField]?.boolValue ?? constraint["default"]?.boolValue ?? false
                    if enabled, let pattern = object[field]?.stringValue {
                        do { _ = try NSRegularExpression(pattern: pattern) }
                        catch { throw invalidTypedPayload("\(path).\(field)", "must be a valid regular expression") }
                    }
                case "stringAnyNonBlank":
                    guard case .array(let fields)? = constraint["fields"] else { continue }
                    let anyNonBlank = fields.compactMap(\.stringValue).contains { field in
                        object[field]?.stringValue?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false
                    }
                    if !anyNonBlank {
                        throw invalidTypedPayload(path, "requires one string field to be non-blank")
                    }
                default:
                    throw ReaderUIRuntimeFailure(code: "INVALID_TYPED_CONTRACT", message: "\(path) has an unsupported constraint")
                }
            }
        }
        return
    case "array":
        guard case .array(let array) = value else { throw invalidTypedPayload(path, "must be an array") }
        if let minimum = schema["minItems"]?.intValue, array.count < minimum {
            throw invalidTypedPayload(path, "must contain at least \(minimum) items")
        }
        if let maximum = schema["maxItems"]?.intValue, array.count > maximum {
            throw invalidTypedPayload(path, "must contain at most \(maximum) items")
        }
        guard let items = schema["items"] else {
            throw ReaderUIRuntimeFailure(code: "INVALID_TYPED_CONTRACT", message: "\(path) array schema has no items")
        }
        for (index, item) in array.enumerated() {
            try validateTypedPayloadSchema(items, value: item, path: "\(path)[\(index)]")
        }
        return
    case "string":
        guard let string = value.stringValue else { throw invalidTypedPayload(path, "must be a string") }
        if let minimum = schema["minLength"]?.intValue, string.count < minimum {
            throw invalidTypedPayload(path, "must contain at least \(minimum) characters")
        }
        if let maximum = schema["maxLength"]?.intValue, string.count > maximum {
            throw invalidTypedPayload(path, "must contain at most \(maximum) characters")
        }
        if schema["nonBlank"]?.boolValue == true,
           string.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            throw invalidTypedPayload(path, "must be non-blank")
        }
        if let pattern = schema["pattern"]?.stringValue {
            let regex: NSRegularExpression
            do { regex = try NSRegularExpression(pattern: pattern) }
            catch {
                throw ReaderUIRuntimeFailure(code: "INVALID_TYPED_CONTRACT", message: "\(path) has an invalid schema pattern")
            }
            let range = NSRange(string.startIndex..<string.endIndex, in: string)
            if regex.firstMatch(in: string, range: range) == nil {
                throw invalidTypedPayload(path, "does not match its required pattern")
            }
        }
    case "integer":
        guard case .number(let number) = value,
              number.isFinite,
              number.rounded(.towardZero) == number,
              abs(number) <= readerUISafeIntegerLimit else {
            throw invalidTypedPayload(path, "must be an IEEE-754 safe integer")
        }
    case "number":
        guard case .number(let number) = value, number.isFinite else {
            throw invalidTypedPayload(path, "must be a finite number")
        }
    case "boolean":
        guard value.boolValue != nil else { throw invalidTypedPayload(path, "must be a boolean") }
    case "json":
        try validateTypedJSONValue(value, path: path)
        return
    default:
        throw ReaderUIRuntimeFailure(code: "INVALID_TYPED_CONTRACT", message: "\(path) has unsupported schema type \(type)")
    }

    if let constant = schema["const"], constant != value {
        throw invalidTypedPayload(path, "does not equal its required constant")
    }
    if case .array(let allowed)? = schema["enum"], !allowed.contains(value) {
        throw invalidTypedPayload(path, "is not an allowed enum value")
    }
    if let minimum = schema["minimum"]?.doubleValue,
       let number = value.doubleValue, number < minimum {
        throw invalidTypedPayload(path, "must be >= \(minimum)")
    }
    if let maximum = schema["maximum"]?.doubleValue,
       let number = value.doubleValue, number > maximum {
        throw invalidTypedPayload(path, "must be <= \(maximum)")
    }
}
